// @larksuiteoapi/api/dist/core/token/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplyAppTicketReq = exports.GetISVAppAccessTokenReq = exports.GetInternalAccessTokenReq = exports.GetISVTenantAccessTokenReq = void 0;
class GetISVTenantAccessTokenReq {
    constructor(app_access_token, tenant_key) {
        this.app_access_token = app_access_token;
        this.tenant_key = tenant_key;
    }
}
exports.GetISVTenantAccessTokenReq = GetISVTenantAccessTokenReq;
class GetInternalAccessTokenReq {
    constructor(app_id, app_secret) {
        this.app_id = app_id;
        this.app_secret = app_secret;
    }
}
exports.GetInternalAccessTokenReq = GetInternalAccessTokenReq;
class GetISVAppAccessTokenReq {
    constructor(app_id, app_secret, app_ticket) {
        this.app_id = app_id;
        this.app_secret = app_secret;
        this.app_ticket = app_ticket;
    }
}
exports.GetISVAppAccessTokenReq = GetISVAppAccessTokenReq;
class ApplyAppTicketReq {
    constructor(app_id, app_secret) {
        this.app_id = app_id;
        this.app_secret = app_secret;
    }
}
exports.ApplyAppTicketReq = ApplyAppTicketReq;
