// @larksuiteoapi/api/dist/core/response/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.newErrorOfInvalidResp = exports.newErr = exports.instanceOfError = exports.ErrCode = void 0;
var ErrCode;
(function (ErrCode) {
    ErrCode[ErrCode["Native"] = -1] = "Native";
    ErrCode[ErrCode["Ok"] = 0] = "Ok";
    ErrCode[ErrCode["AppTicketInvalid"] = 10012] = "AppTicketInvalid";
    ErrCode[ErrCode["AccessTokenInvalid"] = 99991671] = "AccessTokenInvalid";
    ErrCode[ErrCode["AppAccessTokenInvalid"] = 99991664] = "AppAccessTokenInvalid";
    ErrCode[ErrCode["TenantAccessTokenInvalid"] = 99991663] = "TenantAccessTokenInvalid";
    ErrCode[ErrCode["UserAccessTokenInvalid"] = 99991668] = "UserAccessTokenInvalid";
    ErrCode[ErrCode["UserRefreshTokenInvalid"] = 99991669] = "UserRefreshTokenInvalid";
})(ErrCode = exports.ErrCode || (exports.ErrCode = {}));
exports.instanceOfError = (object) => {
    return "code" in object && "msg" in object;
};
exports.newErr = (e) => {
    let err = {
        code: ErrCode.Native,
        msg: e.toString(),
    };
    return err;
};
exports.newErrorOfInvalidResp = (msg) => {
    let err = {
        code: ErrCode.Native,
        msg: msg,
    };
    return err;
};
